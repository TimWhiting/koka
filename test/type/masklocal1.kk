// basic I/O
effect bio
  // Write a string to a file descriptor
  fun write( fd : filedesc, s : string ) : ()

alias filedesc = int

val stdout = 0

fun echo( s : string ) : bio ()
  write(stdout,s)

fun bio( action : () -> <bio|e> a) : e (a,string)
  var buf := ""
  with fun write(fd,s) buf := buf ++ s
  (action(),buf)

fun bio2( action : () -> <bio|e> a) : e (a,string)
  var buf := ""
  with return(x) (x,buf)
  with fun write(fd,s) buf := buf ++ s
  action()

fun bio3( action )
  var buf := ""
  with fun write(fd,s) buf := buf ++ s
  (action(),buf)

fun bio4( action )
  var buf := ""
  with fun write(fd,s) buf := buf ++ s
  (mask<local>(action),buf)

fun bio5( action : () -> <bio|e> a) : e (a,string)
  var buf := ""
  with fun write(fd,s) buf := buf ++ s
  (mask<local>(action),buf)

