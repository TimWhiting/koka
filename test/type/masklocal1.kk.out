type/masklocal1/bio: forall<a,e> (action : () -> <bio|e> a) -> e (a, string)
type/masklocal1/bio2: forall<a,e> (action : () -> <bio|e> a) -> e (a, string)
type/masklocal1/bio3: forall<a,e> (action : () -> <bio|e> a) -> e (a, string)
type/masklocal1/bio4: forall<a,e> (action : () -> <bio|e> a) -> e (a, string)
type/masklocal1/bio5: forall<a,e> (action : () -> <bio|e> a) -> e (a, string)
type/masklocal1/echo: (s : string) -> bio ()
type/masklocal1/stdout: int
type/masklocal1/write: (fd : filedesc, s : string) -> bio ()